astropy/units/quantity.py

```diff
@@ -1084,20 +1084,18 @@ def __ilshift__(self, other):
             return NotImplemented
 
         try:
-            factor = self.unit._to(other)
-        except Exception:
-            # Maybe via equivalencies?  Now we do make a temporary copy.
-            try:
-                value = self._to_value(other)
-            except UnitConversionError:
-                return NotImplemented
-
-            self.view(np.ndarray)[...] = value
-
+            factor = self.unit._to(other)  # except Equivalency
+            self_as_array = self.view(np.ndarray)
+            self_as_array *= factor  # except output dtype
+        # The real error is `numpy.core._exceptions._UFuncOutputCastingError`, which
```

Comment:
It is not clear exactly what the comment refers too -- in principle both the first and third line can error.
Would it be better to do
```
try:
    factor = self.unit._to(other)
except ...:
    return NotImplemented

self_as_array = self.view(np.ndarray)
try:
    self_as_array *= factor
except ...:
    return NotImplemented

self._set_unit...
return self
```

Reply:
This looks good to me.

Reply:
I think ``self_as_array[...] *= factor`` is necessary to preserve the view.

Reply:
I don't think so: `self_as_array` *is* a view. Before the `[...]` was needed because there was a method call; python errors if one does `self.view(np.ndarray) *= factor`.